File: sdk/java/src/test/java/com/anyvali/ConformanceTest.java
package com.anyvali;

import com.anyvali.interchange.Importer;
import com.anyvali.interchange.JsonHelper;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Reads spec/corpus/ JSON files and runs conformance tests.
 */
class ConformanceTest {

    private static final Path CORPUS_DIR = findCorpusDir();

    private static Path findCorpusDir() {
        // Walk up from the SDK directory to find spec/corpus/
        Path current = Paths.get("").toAbsolutePath();
        // Try relative to CWD
        Path candidate = current.resolve("spec/corpus");
        if (Files.isDirectory(candidate)) return candidate;
        // Try from sdk/java
        candidate = current.resolve("../../spec/corpus");
        if (Files.isDirectory(candidate)) return candidate.normalize();
        // Try absolute well-known path
        candidate = Paths.get("C:/Users/MitchellRobert/_win_repos/anyvali/spec/corpus");
        if (Files.isDirectory(candidate)) return candidate;
        // Fallback
        return Paths.get("spec/corpus");
    }

    @TestFactory
    Stream<DynamicTest> conformanceSuite() throws IOException {
        if (!Files.isDirectory(CORPUS_DIR)) {
            return Stream.of(DynamicTest.dynamicTest("corpus dir not found - skipping",
                    () -> System.out.println("Corpus dir not found at " + CORPUS_DIR)));
        }

        List<DynamicTest> tests = new ArrayList<>();

        try (var stream = Files.walk(CORPUS_DIR)) {
            List<Path> jsonFiles = stream
                    .filter(p -> p.toString().endsWith(".json"))
                    .sorted()
                    .toList();

            for (Path file : jsonFiles) {
                String content = Files.readString(file);
                @SuppressWarnings("unchecked")
                Map<String, Object> suite = (Map<String, Object>) JsonHelper.parseJson(content);
                String suiteName = (String) suite.get("suite");
                @SuppressWarnings("unchecked")
                List<Map<String, Object>> cases = (List<Map<String, Object>>) suite.get("cases");

                if (cases == null) continue;

                for (int i = 0; i < cases.size(); i++) {
                    Map<String, Object> testCase = cases.get(i);
                    String description = (String) testCase.get("description");
                    String testName = suiteName + " [" + i + "] " + description;

                    tests.add(DynamicTest.dynamicTest(testName, () -> runTestCase(testCase)));
                }
            }
        }

        return tests.stream();
    }

    @SuppressWarnings("unchecked")
    private void runTestCase(Map<String, Object> testCase) {
        Map<String, Object> schemaDoc = (Map<String, Object>) testCase.get("schema");
        Object input = testCase.get("input");
        boolean expectedValid = (Boolean) testCase.get("valid");
        List<Map<String, Object>> expectedIssues = (List<Map<String, Object>>) testCase.get("issues");

        Schema<?> schema;
        try {
            schema = Importer.importSchema(schemaDoc);
        } catch (Exception e) {
            if (!expectedValid) {
                // Schema import failure on an invalid test case is acceptable
                return;
            }
            fail("Failed to import schema: " + e.getMessage());
            return;
        }

        ParseResult<?> result = schema.safeParse(input);

        assertEquals(expectedValid, result.success(),
                "Expected valid=" + expectedValid + " but got " + result.success()
                + (result.issues().isEmpty() ? "" : " issues: " + result.issues()));

        if (!expectedValid && expectedIssues != null && !expectedIssues.isEmpty()) {
            // Verify issue codes match
            assertEquals(expectedIssues.size(), result.issues().size(),
                    "Expected " + expectedIssues.size() + " issues but got " + result.issues().size()
                    + ": " + result.issues());

            for (int i = 0; i < expectedIssues.size(); i++) {
                Map<String, Object> expectedIssue = expectedIssues.get(i);
                ValidationIssue actualIssue = result.issues().get(i);

                // Check code
                String expectedCode = (String) expectedIssue.get("code");
                if (expectedCode != null) {
                    assertEquals(expectedCode, actualIssue.code(),
                            "Issue code mismatch at index " + i);
                }

                // Check path
                List<Object> expectedPath = (List<Object>) expectedIssue.get("path");
                if (expectedPath != null) {
                    assertEquals(expectedPath.size(), actualIssue.path().size(),
                            "Path size mismatch at issue index " + i);
                    for (int j = 0; j < expectedPath.size(); j++) {
                        Object expectedPathElem = expectedPath.get(j);
                        Object actualPathElem = actualIssue.path().get(j);
                        // Compare as strings for flexibility (int vs string path segments)
                        assertEquals(String.valueOf(expectedPathElem),
                                String.valueOf(actualPathElem),
                                "Path element mismatch at issue " + i + " path index " + j);
                    }
                }
            }
        }

        // Verify output on successful parse
        if (expectedValid && testCase.containsKey("output")) {
            Object expectedOutput = testCase.get("output");
            assertOutputMatches(expectedOutput, result.data());
        }
    }

    private void assertOutputMatches(Object expected, Object actual) {
        if (expected == null) {
            assertNull(actual);
            return;
        }
        if (expected instanceof Map<?, ?> expectedMap && actual instanceof Map<?, ?> actualMap) {
            assertEquals(expectedMap.size(), actualMap.size(),
                    "Map size mismatch: expected " + expectedMap + " but got " + actualMap);
            for (var entry : expectedMap.entrySet()) {
                assertTrue(actualMap.containsKey(entry.getKey()),
                        "Missing key: " + entry.getKey());
                assertOutputMatches(entry.getValue(), actualMap.get(entry.getKey()));
            }
        } else if (expected instanceof List<?> expectedList && actual instanceof List<?> actualList) {
            assertEquals(expectedList.size(), actualList.size(),
                    "List size mismatch");
            for (int i = 0; i < expectedList.size(); i++) {
                assertOutputMatches(expectedList.get(i), actualList.get(i));
            }
        } else if (expected instanceof Number && actual instanceof Number) {
            // Compare numeric values flexibly (int/long/double)
            assertEquals(((Number) expected).doubleValue(), ((Number) actual).doubleValue(), 1e-9,
                    "Numeric mismatch: expected " + expected + " but got " + actual);
        } else {
            assertEquals(expected, actual);
        }
    }
}

File: sdk/kotlin/src/test/kotlin/com/anyvali/ConformanceTest.kt
package com.anyvali

import com.anyvali.interchange.Importer
import kotlinx.serialization.json.*
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertTrue

class ConformanceTest {

    private val json = Json { ignoreUnknownKeys = true }

    private fun findCorpusDir(): File {
        // Navigate from sdk/kotlin up to repo root, then into spec/corpus
        var dir = File(System.getProperty("user.dir"))
        // If we're in sdk/kotlin, go up two levels
        val specCorpus = File(dir, "../../spec/corpus")
        if (specCorpus.exists()) return specCorpus.canonicalFile

        // Try from repo root
        val direct = File(dir, "spec/corpus")
        if (direct.exists()) return direct.canonicalFile

        throw IllegalStateException("Cannot find spec/corpus directory from $dir")
    }

    private fun loadTestSuites(): List<Pair<String, JsonObject>> {
        val corpusDir = findCorpusDir()
        val suites = mutableListOf<Pair<String, JsonObject>>()
        corpusDir.walkTopDown()
            .filter { it.isFile && it.extension == "json" }
            .forEach { file ->
                val content = file.readText()
                val obj = json.parseToJsonElement(content).jsonObject
                suites.add(file.nameWithoutExtension to obj)
            }
        return suites
    }

    @TestFactory
    fun `conformance test corpus`(): List<DynamicTest> {
        val suites = loadTestSuites()
        val tests = mutableListOf<DynamicTest>()

        for ((suiteName, suiteObj) in suites) {
            val cases = suiteObj["cases"]?.jsonArray ?: continue

            for (caseEl in cases) {
                val caseObj = caseEl.jsonObject
                val description = caseObj["description"]?.jsonPrimitive?.content ?: "unknown"
                val schemaDoc = caseObj["schema"]!!.jsonObject
                val inputEl = caseObj["input"]!!
                val valid = caseObj["valid"]!!.jsonPrimitive.boolean
                val expectedIssues = caseObj["issues"]?.jsonArray ?: JsonArray(emptyList())

                tests.add(DynamicTest.dynamicTest("$suiteName: $description") {
                    runConformanceCase(schemaDoc, inputEl, valid, expectedIssues)
                })
            }
        }

        assertTrue(tests.isNotEmpty(), "Should have found at least one test case")
        return tests
    }

    private fun runConformanceCase(
        schemaDoc: JsonObject,
        inputEl: JsonElement,
        expectedValid: Boolean,
        expectedIssues: JsonArray
    ) {
        val (schema, definitions) = Importer.importFromJsonObject(schemaDoc)
        val input = jsonElementToKotlin(inputEl)

        val result = schema.safeParse(input, definitions)

        assertEquals(
            expectedValid,
            result.isSuccess,
            "Expected valid=$expectedValid but got ${if (result.isSuccess) "success" else "failure: ${(result as ParseResult.Failure).issues}"}"
        )

        if (!expectedValid) {
            assertIs<ParseResult.Failure>(result)
            // Check that we got the expected issue codes
            for ((index, expectedIssueEl) in expectedIssues.withIndex()) {
                val expectedIssue = expectedIssueEl.jsonObject
                val expectedCode = expectedIssue["code"]!!.jsonPrimitive.content
                if (index < result.issues.size) {
                    assertEquals(
                        expectedCode,
                        result.issues[index].code,
                        "Issue[$index] code mismatch"
                    )
                    // Check path
                    val expectedPath = expectedIssue["path"]!!.jsonArray.map { pathEl ->
                        when {
                            pathEl is JsonPrimitive && pathEl.isString -> pathEl.content
                            pathEl is JsonPrimitive -> pathEl.int
                            else -> pathEl.toString()
                        }
                    }
                    assertEquals(
                        expectedPath,
                        result.issues[index].path,
                        "Issue[$index] path mismatch"
                    )
                }
            }
            assertEquals(
                expectedIssues.size,
                result.issues.size,
                "Expected ${expectedIssues.size} issues but got ${result.issues.size}: ${result.issues}"
            )
        }

        if (expectedValid) {
            assertIs<ParseResult.Success>(result)
        }
    }

    private fun jsonElementToKotlin(element: JsonElement): Any? {
        return when (element) {
            is JsonNull -> null
            is JsonPrimitive -> {
                if (element.isString) {
                    element.content
                } else {
                    val content = element.content
                    if (content == "true") return true
                    if (content == "false") return false
                    // Try long first, then double
                    content.toLongOrNull()?.let { return it }
                    content.toDoubleOrNull()?.let { return it }
                    content
                }
            }
            is JsonArray -> element.map { jsonElementToKotlin(it) }
            is JsonObject -> element.entries.associate { (k, v) -> k to jsonElementToKotlin(v) }
        }
    }
}
